File: 139-word-break/word-break.cpp
class Solution {
public:
    int dp[301];
    unordered_set<string> set;
    int solve(string &s, int start)
    {
        if(start == s.length())
        return true;

        if(dp[start] != -1)
        return dp[start];

        string curr="";
        for(int i = start; i < s.length(); i++)
        {
            curr.push_back(s[i]);
            if(set.count(curr) && solve(s,i+1))
            return 1;
        }
        // return false
        return dp[start]=0;
    }

    bool wordBreak(string s, vector<string>& wordDict) 
    {
        memset(dp,-1,sizeof(dp));

        for(string word : wordDict)
        set.insert(word);
        
        return solve(s,0);
    }
};

File: 139-word-break/word-break.java
class Solution {
    private Set<String> wordSet;
    private Map<String, Boolean> dp;
    private boolean check(String word) {
        if(wordSet.contains(word)) return true;
        if(dp.containsKey(word)) return dp.get(word);
        for(int i = 0; i < word.length(); i++){
            String prefix = word.substring(0, i + 1);
            String suffix = word.substring(i + 1);
            if(wordSet.contains(prefix) && check(suffix)){
                dp.put(word, true);
                return true;
            }
        }
        dp.put(word, false);
        return false;
    }
    public boolean wordBreak(String s, List<String> wordDict) {
        wordSet = new HashSet<>(wordDict);
        dp = new HashMap<>();
        return check(s);
    }
}